src/Compilers/CSharp/Portable/Symbols/Source/SourceLabelSymbol.cs

```diff
@@ -17,21 +17,48 @@ internal sealed class SourceLabelSymbol : LabelSymbol
         /// </summary>
         private readonly ConstantValue _switchCaseLabelConstant;
 
+        // PERF: Often we do not need this, so we make this lazy
+        private string _lazyName;
+
         public SourceLabelSymbol(
             MethodSymbol containingMethod,
             SyntaxNodeOrToken identifierNodeOrToken,
             ConstantValue switchCaseLabelConstant = null)
-            : base(identifierNodeOrToken.IsToken ? identifierNodeOrToken.AsToken().ValueText : identifierNodeOrToken.ToString())
         {
             _containingMethod = containingMethod;
             _identifierNodeOrToken = identifierNodeOrToken;
             _switchCaseLabelConstant = switchCaseLabelConstant;
         }
 
+        public override string Name
+        {
+            get
+            {
+                return _lazyName ??
+                    (_lazyName = MakeLabelName());
+            }
+        }
+
+        private string MakeLabelName()
+        {
+            var node = _identifierNodeOrToken.AsNode();
```

Comment:
The node test is cheaper. 
AsNode returns a single reference - node or null, where AsToken returns a struct with embedded references and caller needs to check its kind to figure if it actually succeeded.

AsToken check is typically done first in cases where it is way more likely to succeed than AsNode. I am not sure it is one of those cases.